src/nucleus/sanity.ml

```diff
@@ -76,25 +76,35 @@ let rec type_of_term_abstraction sgn = function
      let t_abstr = type_of_term_abstraction sgn abstr in
      Abstract.abstraction Abstract.is_type a.atom_nonce t_abstr
 
-let rec abstracted_judgement_with_boundary sgn abstr_jdg =
-  match abstr_jdg with
-    | NotAbstract (JudgementIsTerm jdg as jdg') ->
-      Mk.not_abstract (jdg', BoundaryIsTerm (type_of_term sgn jdg))
+let judgement_with_boundary sgn jdg =
+  match jdg with
+    | JudgementIsTerm jdg as jdg'->
+      jdg', BoundaryIsTerm (type_of_term sgn jdg)
 
-    | NotAbstract (JudgementIsType jdg as jdg') ->
-      Mk.not_abstract (jdg', BoundaryIsType ())
+    | JudgementIsType jdg as jdg'->
+      jdg', BoundaryIsType ()
 
-    | NotAbstract (JudgementEqType (EqType (asmp, t1, t2)) as jdg') ->
-      Mk.not_abstract (jdg', BoundaryEqType (t1,t2))
+    | JudgementEqType (EqType (asmp, t1, t2)) as jdg'->
+      jdg', BoundaryEqType (t1,t2)
 
-    | NotAbstract (JudgementEqTerm (EqTerm (asmp, e1, e2, t)) as jdg')->
-      Mk.not_abstract (jdg', BoundaryEqTerm (e1, e2, t))
+    | JudgementEqTerm (EqTerm (asmp, e1, e2, t)) as jdg'->
+      jdg', BoundaryEqTerm (e1, e2, t)
 
+let rec abstracted_judgement_with_boundary sgn abstr_jdg =
+  match abstr_jdg with
     | Abstract (x, t, abstr) ->
       let a, abstr = Unabstract.abstraction Instantiate_bound.judgement x t abstr in
       let abstr_pair = abstracted_judgement_with_boundary sgn abstr in
       Abstract.judgement_with_boundary_abstraction a abstr_pair
 
+    | NotAbstract jdg ->
+      Mk.not_abstract (judgement_with_boundary sgn jdg)
+
+let rec derivation_with_boundary sgn der =
+  match der with
+    | Premise (mv, der) -> Premise (mv, derivation_with_boundary sgn der)
```

Comment:
Here's the bug. We should never go under a meta-variable binder because we'll encounter a bound meta-variable (and we do).